CSIS API update

Thalley opened the pull request:
Update the API for the CSIS server and client.

fixes https://github.com/zephyrproject-rtos/zephyr/issues/41591

Thalley committed:
Bluetooth: CSIS: Move csis_instance to csis_internal.h

Move the struct definition to the internal header file,
and add it to the bt_csis struct.

This also renames the csis_instance to
bt_csis_client_svc_inst to use the bt_csis prefix.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Rename servers to client_insts in csis_client.c

Rename the array to a more descriptive name and change the type
to bt_csis, as well as renaming the individual variables
when accessing the array.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Use bt_csis for CSIS client

Use the bt_csis struct instead of the
bt_csis_client_svc_inst struct for the CSIS client.
This makes it more similar to not only the CSIS
implementation, but also the other LE Audio
client implementations.

Furthermore, this change will make it easier to use
bt_csis in the API in the future.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Add array of bt_csis to discovery function

Add an [out] array to the discovery function and
an array to the discovery callback of bt_csis structs.
These should be used instead of bt_conn pointers for
CSIS client.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Expose bt_csis to client via bt_csis_client_set_member

Use the bt_csis_client_set_member struct to store the individual
bt_csis client struct. This way they are exposed to the
client application.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Remove unused functions and callbacks from csis.h

Remove the lock and release sets functions, as well
as the discover member function as they have been removed
from the implementation a while ago.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Refactor bt_csis_client_discover_sets to use member

Refactor bt_csis_client_discover_sets to use the
bt_csis_client_set_member struct instead of a bt_conn.
The bt_csis_client_set_member represents a remote server
(set member), and make it possible to avoid sending indexes
of instances around instead of bt_csis.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Remove conn from bt_csis_client_lock_changed_cb

The connection pointer can be inferred based on the
set pointer, by using e.g. CONTAINER_OF.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Refactor bt_csis_client_lock_read_cb to use set

Refactor the bt_csis_client_lock_read_cb callback to use
a bt_csis_client_set pointer instead of conn and index.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Fix bad lock and release set error values

Some lock_set and release_set callback had incorrect error
messages.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Refactor bt_csis_client_lock_get to use member

Refactor the bt_csis_client_lock_get function to use a
pointer to a member and a set instead of a bt_conn
pointer and an index.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Remove BT_CSIS_ERROR_SIRK_ACCESS_REJECTED

Remove the error code BT_CSIS_ERROR_SIRK_ACCESS_REJECTED
as it no longer exists in the specification.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Don't expose SIRK type to upper layers

Change how the SIRK is exposed to the upper layers.
The SIRK will always be the unencrypted 16 octet
SIRK now, instead of a struct.

This not only allows us to avoid having a
__packed struct in the API, but also gives a better
API as we don't expose encrypted data to the upper layers.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Remove addr from bt_csis_client_set_member

Remove the addr struct from bt_csis_client_set_member as that
was only used by the upper layers and not the CSIS client
itself, and as such should only reside in the
upper layers.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Document missing functions in csis.h

Document the functions and callbacks, macros, etc. that
was missing in the file.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

MariuszSkamra commented on include/bluetooth/audio/csis.h:
```diff
@@ -292,14 +292,15 @@ void bt_csis_client_register_cb(struct bt_csis_client_cb *cb);
  * @return Return 0 on success, or an errno value on error.
  */
-int bt_csis_client_lock_get(struct bt_conn *conn, uint8_t inst_idx);
+int bt_csis_client_lock_get(struct bt_csis_client_set_member *member,
+			    const struct bt_csis_client_set *set);
```
Two minor comments or rather questions from me:
1. First of all, do the API user needs to read the lock state by itself? Is this function needed to be exposed in public API? Shouldn't the user call to acquire and release the lock if needed only?
2. Do you think the function name suffix *_get* reminds *_acquire*? But it's not what this function does actually. Maybe you could consider changing name to *_get_state*.

Thalley replied:
1. It's a good question, and the answer is "maybe not". A user may want to ensure that all set members are unlocked before attempting to lock the set (as recovering from a failed set lock procedure is more expensive than reading the lock state on all devices first. Furthermore, we will need this to pass qualification tests iirc. 

2. I had never considered that `get` may be misunderstood in this case, but I definitely see your point. `get_state` is much better.

MariuszSkamra replied:
Ad. 1. But isn't the CSIS implementation job to track the member lock state? I guess it should do it, because it subscribes to Set Member Lock ATT Notification. So the CSIS implementation knows a priori whether member is locked and whether it's locked by us before user calls API bt_csis_client_lock/bt_csis_client_release.
As API user I would expect the CSIS to do acquire/release when requested and/or return success or error on failure.
If the logic will be done in this service, one would not need to handle it in upper layers e.g. CAP.

Regarding qualification test, if there is a test case that requires ATT Read Request to Lock State we could pass it anyway if the stack will read the Lock State right after connection. It has to be done anyway IMO knowing the state, implementation decides whether to request or not the Lock if it's already granted to someone else.

Ad. 2 Thanks ^^

Thalley replied:
Assuming that we read the lock state value when connect to a non-bonded device, then we would still need to do the discovery first. It isn't something that we have considered much before, but we don't usually do discovery on connection established autonomously in Zephyr, which would be required to read the lock state. Due to that dependency, we can't auto-read the state (although we could read it as part of discovery procedure).
For bonded devices, we should expect a notification on reconnection. 
We will also need to assume that notifications are not lost on air (or otherwise), if we don't want to expose an explicitly read procedures. 

If we need to implement the functionality in the first place, it could also be argued that it doesn't hurt to expose this as an API.  

So the way I see it, we have the options of
1) Auto discover and auto read lock state on connection
2) Read lock state during discovery procedure
3) Expose the get_lock_state function

Where 1) is a major change, 3) is what we have and 2) can fairly easily be implemented.

MariuszSkamra replied:
I know that Zephyr don't discover ATT attributes after ACL is established autonomously.
I was thinking rather that the Lock State read could be part if bt_csis_client_discover (sorry I was unclear).

Thalley replied:
Given that there isn't a defined procedure for it in the spec, I think it would be fine to keep in the discovery. 

Will update this PR with a new commit that changes that.

asbjornsabo replied:
I will have to read the whole PR to get the full picture, but here are some inputs on this point.
- Doing discovery autonomously and directly after connection could be a pretty large load on the system in the cases where there are many services (as there may be in le-audio).  I believe some applications would rather prefer to be in control of this.
- For the lock characteristic, notifications are mandatory, but for the other characteristics of CSIS, notifications are not mandatory.  So, leaving out the read (API) for the lock would make it different than the other characteristics.  (Which may be OK, or may be seen as lack of consistency.)
- Even if the notifications are mandatory, notifications are not reliable (over ATT).  So the state kept may be wrong.  (Not an issue over EATT, but we have had as a policy to not rely upon, or mandate, EATT.
- Since the "get" was mentioned: "set" is also open for interpretation - does lock_set() lock the set, or does it set the lock? (Not really a problem in that case, as locking the set implies setting the lock ...)

rymanluk replied:
> * Even if the notifications are mandatory, notifications are not reliable (over ATT).  So the state kept may be wrong.
could you point me to the 

Could you please point out the place in the Zephyr Bluetooth code, where Notification packet could be dropped, just because this is notification or because of the lack of the buffers?
I have feeling this is not a case for the Zephyr. Is so then saying "notification as not reliable" is incorrect.

rymanluk replied:
> * Doing discovery autonomously and directly after connection could be a pretty large load on the system in the cases where there are many services (as there may be in le-audio).  I believe some applications would rather prefer to be in control of this.

@asbjornsabo What would be policy for the discovery start for such an applications if not the Connection Establishment time? I'm curious :)

Thalley replied:
> Could you please point out the place in the Zephyr Bluetooth code, where Notification packet could be dropped, just because this is notification or because of the lack of the buffers?

We don't do that in Zephyr AFAIK, but that doesn't mean other implementations won't. 

> What would be policy for the discovery start for such an applications if not the Connection Establishment time? I'm curious :)

Application specific I suppose, of if requirement from other profiles :) 

LE Audio is a bit unique in this regard, where both the central and peripheral need to do discovery, as otherwise that has mostly been the central, so it would not make sense to for a Zephyr device acting as a peripheral to always do a discovery.

rymanluk replied:
> We don't do that in Zephyr AFAIK, but that doesn't mean other implementations won't.
> 
> >

Agree, then we are good. As a Client we can rely on the notifications because those are reliable on Zephyr.
And as a Server we will follow what other Client expects as to do and leave that issue to solve to other implementations.

Thalley replied:
@rymanluk Right, I think I finally understood what you mean. 

Since the dropping of notifications is a GATT client only issue, then the server won't care about that. 
The Zephyr client will never drop notifications due to buffer issues (perhaps it should?), but Zephyr does not need this procedure to function properly as a set coordinator. 

The point is moot though, as the procedure is part of the CAP test spec (details not public), so we will need to perform this procedure to qualify CAP, unless we change the CAP test spec. 

I suggest that if this is the only blocker for this PR, that we get it merged and then if we can avoid the procedure all-together due to the specification changes, then we can remove it for optimization. Does that sound OK to you?

rymanluk replied:
@Thalley  That is fine, however I would not worry about CAP Test spec - because the only thing we need to do is read Lock, and this we need to do anyway after first connection.

Thalley replied:
@rymanluk But if the test have the sequence:

1) CSIP discover (discovery + read lock)
2) CAP discover
3) CSIP Read Ordered Access
4) Do some other procedure 
(which is the case in some CAP tests)

Then we would otherwise need to do csip_discover() for step 3 again to do the read of the lock. So we could probably hack our way into passing the test, but we'd need to do discovery do perform a lock read if we don't have this procedure implemented.

Thalley committed:
Bluetooth: CSIS: Rename bt_csis_client_lock_get

Rename bt_csis_client_lock_get to
bt_csis_client_get_lock_state. `get` could be
misunderstood as acquire, i.e. that `get` would
mean that the lock was taken by this device.

The new name should make it more obviously that it
is just a read procedure.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Fix wrong handle check in csis_client_write_set_lock

The wrong handle value was checked.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Remove set handle check from verify_members_and_get_inst

The verify_members_and_get_inst can be used for more than just
locking the set, so checking the specific handle in that
does not make sense.

The handle is furthermore already checked in
csis_client_write_set_lock.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Modify bt_csis_client_get_lock_state to read all members

Modify bt_csis_client_get_lock_state to be the Ordered Access
procedure, which means that instead of reading a single lock value
on a single device, it will read the lock value for all
set members supplied in the function, and return true if any
of them is locked, or false otherwise.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

rymanluk commented on subsys/bluetooth/audio/csis_client.c:
```diff
@@ -1285,6 +1310,37 @@ static int verify_members_and_get_inst(const struct bt_csis_client_set_member **
+		return err;
+	}
+
+	err = csis_client_read_set_lock(cur_inst);
```
do we need to read it always? Don't we register for the notifications and maintain the current lock state?

So in fact, we need to read it once, after bonding, and then we should rely on notifications if we are bonded.
If not bonded, that read it only once after reconnection, and then rely on the notifications.

asbjornsabo replied:
Notifications on ATT are not reliable.  It seems to follow that one should not rely upon them ;-)

Thalley replied:
> Notifications on ATT are not reliable. It seems to follow that one should not rely upon them ;-)

**unreliable** is really the keyword to consider when determining if you should **rely** on it indeed :D

asbjornsabo commented on include/bluetooth/audio/csis.h:
```diff
@@ -193,121 +193,135 @@ int bt_csis_advertise(struct bt_csis *csis, bool enable);
+	 * @brief The 16 octet set Set Identity Resolving Key (SIRK)
+	 *
+	 * The Set SIRK may not be exposed by the server over Bluetooth, and
+	 * may required an out-of-band solution.
```
required -> require

Thalley committed:
Bluetooth: CSIS: Update minor documentation for the CSIS server

A few of the functions were poorly or wrongly described.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo commented on include/bluetooth/audio/csis.h:
```diff
@@ -193,121 +193,135 @@ int bt_csis_advertise(struct bt_csis *csis, bool enable);
-
-typedef void (*bt_csis_client_discover_members_cb)(int err, uint8_t set_size,
-						   uint8_t members_found);
+int bt_csis_client_discover_sets(struct bt_csis_client_set_member *member);
```
The difference between `discover()` and `discover_sets()` is not fully clear to me.
I believe the first one is to be called first, but they take the same input, and their callbacks return the same values.

Thalley replied:
Agreed, they are very similar, and actually both of them are needed for the discovery procedure. 

The first, `discover()` discover the CSIS instances, if any, and initializes the handles. 
The second `discover_sets()` continues the discovery procedure by reading the SIRK, size and rank (if exposed). 

We could consider merging these two, as they are the same procedure. It was intentionally split into two in case that the device did not want to read all the characteristic during discovery, but maybe we need to revisit that idea. Merging the two may be needed to be fully compliant, as a set coordinator _shall_ read the characteristics during the discovery procedure. 

So the first discovers the CSIS service and the GATT handles, the second continues the discovery procedure which includes reading the characteristics. 

**So should we merge the two into a single function (which is more complex), or just rename this one into something better (suggestions are welcome)?**

asbjornsabo replied:
I vote to keep them separate, but update name and documentation.
Proposal:
"discover()" -> "discover_service()", with the documentation saying that
- this is the first discovery function to be run, and that it discovers the existence of the service, but not the characteristics of the service
"discover_sets()" keeps its name, documentation extended to say that it is run after "discover_service()", and discover sets by discovering SIRK, size and rank.

Thalley replied:
> "discover()" -> "discover_service()", with the documentation saying that

I am strongly against this one, unless we update it for the other clients as well. 

How about we rename `discover_sets` to `read_sets` which is more aligned to what it actually does (read the set characteristics). 

If we keep them separate, `discover()` will not be compliant with the discover procedure (Coordinated Set Discovery procedure) which is a good reason to merge them.

asbjornsabo replied:
> > "discover()" -> "discover_service()", with the documentation saying that
> 
> I am strongly against this one, unless we update it for the other clients as well.

Consistency - good point.  I agree.

> How about we rename `discover_sets` to `read_sets` which is more aligned to what it actually does (read the set characteristics).

Seems fine.  I still think the documentation of both could be updated with a little bit of your explanation above.
> 
> If we keep them separate, `discover()` will not be compliant with the discover procedure (Coordinated Set Discovery procedure) which is a good reason to merge them.

You have convinced me.  Having a single function to do the procedure seems cleaner.  And I do not really see now why one would want to find the services, but not their content.  So merge them.

Thalley replied:
> And I do not really see now why one would want to find the services, but not their content.

That is more typical for the rest of the LE Audio services though. Sometimes you want to discover the service, and only read specific values during discovery (unlike CSIP which requests to read _all_). 

I'll merge the two functions and update this PR.

asbjornsabo commented on include/bluetooth/audio/csis.h:
```diff
@@ -345,15 +343,21 @@ int bt_csis_client_release_set(void);
+				  const struct bt_csis_client_set *set);
 
 /**
  * @brief Lock an array of set members
```
This may be a suitable place to try to explain the problems I have with naming.
To me, an "array of set members" (like it is used in this function) sounds like a set (a number of devices that belong together).
However, the set members each contain an "array of sets", and a set (of that kind) is really a "set membership" (?)
I'll probably get it with time, but I think it is possible to find better naming.

Could the lowest level (bt_csis_client_set) be called e.g. or "set_info" to indicate that this is information about a set the device is a member of (rather than a set)?
The next level (_set_member) could be kept as is (I believe this is the information for one device, containing information about the device is a member of).
The top level (array of set members) could then be called a set.

asbjornsabo replied:
Or possibly the lowest lever could be a "set_membership"?
Then a set would be an array of "set_member"s, which would each have a number of "set_memberships".

Thalley replied:
> It would also help if we could (here) use "set" only for the noun, and not for the verb.

Yeah, it would be best to avoid any form of `set_set` :D 

> To me, an "array of set members" (like it is used in this function) sounds like a set (a number of devices that belong together).

They are similar, but not the same. 
A set may consist of 3 devices {SM_1, SM_2, SM_3}, but an array of set members could be a single set member [SM_1] or some [SM_1, SM_2] or all [SM_1, SM_2, SM_3]. 

Likewise, there may also be two sets: {SM_1, SM_2} and {SM_2, SM_3}, i.e. SM_2 is part of multiple sets, and as such must have multiple CSIS instances.  

I agree that `bt_csis_client_set` is not the optimal naming, as it really a representation of a remote CSIS instance, and not a set.  `bt_csis_client_set_info` seem to be a better choice indeed, but it is also contains more than information, so that is not optimal either in my opinion. Perhaps something like `bt_csis_client_inst`?

asbjornsabo replied:
I see your point that the array may be only a partial set.  So it may not be correct to call the array for a set.

As for the second point, I see that my comment was lacking some words - I agree that at set_member (?) may be a member of several sets.

Your proposal for naming (`bt_csis_client_inst`) is an improvement over the current naming - I would be satisfied with that.

asbjornsabo replied:
I can live with that - we have seen worse before.  I find it better than calling the CSIS-instance a "set".

TheCSIS instance contains the SIRK, the set size and the rank of a device (plus an internal member), and there is one such instance for each set the device is a member of.  It could also possibly be called a "membership".  (In which case a set member will end up with several memberships, which seems OK.)

Thalley replied:
I'll update the PR with `bt_csis_client_csis_inst`. 

If we ever rename the API to follow the roles instead, it would become something like `bt_set_coordinator_csis_inst` or possibly `bt_set_coord_csis_inst` which would avoid the douple `csis`.

asbjornsabo reviewed (commented):
I am not totally happy about the naming  - what is called a "set".
Once that is resolved (one way or another) I think this can go in.

Thalley committed:
Bluetooth: CSIS: Add bt_csis_client_set_info struct

Several APIs worked on the bt_csis_client_set struct,
which not only included information about a set, but
also a reference to a specific CSIS instance.

A specialized struct only for the set information
is more useful in those scenarios.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Rename bt_csis_client_set

Rename struct bt_csis_client_set to
struct bt_csis_client_csis_inst, as that is more descriptive
of the actual content of the struct.

This also avoids the confusion about what a "set" is,
which is clearly not a single instance of CSIS
on a single remote server.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: CSIS: Merge the two client discovery functions

Merges bt_csis_client_discover and
bt_csis_client_discover_sets, as they should be done
together for the discovery procedure from the CSIP
spec.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
@asbjornsabo For the naming, I've introduce a new struct `bt_csis_client_set_info` and renamed the `bt_csis_client_set` to `bt_csis_client_csis_inst`. 

I've also merge the two discovery functions. 

All the changes have been done in newly added commits to this PR.

asbjornsabo reviewed (approved):
I am fine with this.  A good change, I think!

Thalley commented:
@rymanluk @asbjornsabo  The only remaining part of this PR is now whether or not to keep the `bt_csis_client_get_lock_state` function, which implements one of the procedures in the CSIP spec, and is used by CAP. 

The lock state is read during discovery and we subscribe to notifications, so there's no need for the application to use it (unless notifications go missing). 

Personally I am in favor of keep the function, and then let the upper layers decide whether to use it or not.

rymanluk reviewed (approved):
Looks good to me

Thalley commented:
@jhedberg @alwa-nordic Any chance you guys can take a look at this, so we have a chance to get it merged before Friday?
